File: y2024/day06/day06.py
import os
import time
from copy import deepcopy
from enum import Enum, auto
from pprint import pprint
from typing import Literal


class Directions(Enum):
    east = auto()
    west = auto()
    north = auto()
    south = auto()


POSITIONS = """....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#..."""


def process_input():
    with open("./day06/input.txt", "r") as fp:
        content = fp.readlines()
    content = POSITIONS.splitlines()
    content = [list(line) for line in content]
    # pprint(content)
    return content


def get_guard_position_orientation(
    positions: list[list[str]],
) -> tuple[tuple[int, int], Directions]:
    for index, row in enumerate(positions):
        try:
            col = row.index("^")
            return (index, col), Directions.north
        except:
            continue
    raise ValueError("Guard not found!")


def calculate_next_pos(
    curr_pos: tuple[int, int], direction: Directions
) -> tuple[int, int]:
    if direction == Directions.east:
        next_pos = curr_pos[0], curr_pos[1] + 1
    elif direction == Directions.west:
        next_pos = curr_pos[0], curr_pos[1] - 1
    elif direction == Directions.north:
        next_pos = curr_pos[0] - 1, curr_pos[1]
    else:
        next_pos = curr_pos[0] + 1, curr_pos[1]
    return next_pos


def reorient(direction: Directions) -> tuple[tuple[int, int], Directions]:
    if direction == Directions.north:
        direction = Directions.east
    elif direction == Directions.east:
        direction = Directions.south
    elif direction == Directions.south:
        direction = Directions.west
    else:
        direction = Directions.north
    return direction


def get_tick(direction: Directions) -> str:
    if direction == Directions.north:
        return "^"
    elif direction == Directions.east:
        return ">"
    elif direction == Directions.south:
        return "v"
    else:
        return "<"


def print_pos(positions: list[list[str]]) -> None:
    pos = ["".join(i) for i in positions]
    pos = "\n".join(pos)
    print(pos)


def move(
    positions: list[list[str]],
    direction: Directions,
    curr_pos: tuple[int, int],
) -> dict[tuple[int, int], Literal[True]]:
    moved_positions = {}
    rows, columns = len(positions), len(positions[0])
    next_pos = curr_pos
    while True:
        # tick = get_tick(direction)
        # positions[curr_pos[0]][curr_pos[1]] = tick
        # os.system("clear")
        # print_pos(positions)

        moved_positions[curr_pos] = True
        next_pos = calculate_next_pos(curr_pos, direction)
        if (
            next_pos[0] >= rows
            or next_pos[0] < 0
            or next_pos[1] >= columns
            or next_pos[1] < 0
        ):
            break
        if positions[next_pos[0]][next_pos[1]] == "#":
            direction = reorient(direction)
            continue
        curr_pos = next_pos
        # time.sleep(0.25)
    return moved_positions


def move_loop(
    positions: list[list[str]], direction: Directions, curr_pos: tuple[int, int]
) -> True:
    moved_positions: dict[tuple[int, int, int], Literal[True]] = {}
    rows, columns = len(positions), len(positions[0])
    next_pos = curr_pos
    while True:
        # tick = get_tick(direction)
        # positions[curr_pos[0]][curr_pos[1]] = tick
        # os.system("clear")
        # print_pos(positions)
        mp_key = (*curr_pos, direction.name)
        if moved_positions.get(mp_key, None):
            moved_positions[mp_key] += 1
            return True
        else:
            moved_positions[mp_key] = 1
        next_pos = calculate_next_pos(curr_pos, direction)
        if (
            next_pos[0] >= rows
            or next_pos[0] < 0
            or next_pos[1] >= columns
            or next_pos[1] < 0
        ):
            return False
        if positions[next_pos[0]][next_pos[1]] == "#":
            direction = reorient(direction)
            continue
        curr_pos = next_pos
        check = [v for _, v in moved_positions.items() if v > 3]
        if sum(check) > len(moved_positions) * 5:
            print(f"loop identified, all position traversed at least once!")
            return True
    return moved_positions


def part01(positions) -> dict[tuple[int, int], Literal[True]]:
    pos, direction = get_guard_position_orientation(positions)
    moved_positions = move(positions, direction, pos)
    print("Part 01:", len(moved_positions))
    # not at start
    return moved_positions


def part02(positions: list[list[str]]):
    moved_positions = part01(deepcopy(positions))
    pos, direction = get_guard_position_orientation(positions)
    loop_count = 0
    pkeys = list(moved_positions.keys())
    perv_pos_index = 0
    for _, k in enumerate(pkeys[1:]):
        positions[k[0]][k[1]] = "#"
        is_loop = move_loop(positions, direction, pos)
        positions[k[0]][k[1]] = "."
        perv_pos_index += 1
        if is_loop:
            loop_count += 1


def main():
    data = process_input()
    part02(data)


if __name__ == "__main__":
    main()

File: y2024/day06/day06.go
package day06

import (
	"adventofcode/y2024/utils"
	"crypto/sha256"
	"fmt"
	"log"
)

var POSITIONS = []string{
	"....#.....",
	".........#",
	"..........",
	"..#.......",
	".......#..",
	"..........",
	".#..^.....",
	"........#.",
	"#.........",
	"......#...",
}

const (
	east  = iota
	west  = iota
	north = iota
	south = iota
)

type Position struct {
	Row   int
	Col   int
	Value int
}

func (p *Position) PosHashKey() string {
	h := sha256.New()
	h.Write([]byte(fmt.Sprintf("(%d,%d)", p.Row, p.Col)))
	return string(h.Sum(nil))
}

func processInput(dayDir string) [][]string {
	lines := utils.ReadLines(dayDir + "/input.txt")
	retval := [][]string{}
	for _, line := range lines {
		row := []string{}
		for _, ch := range line {
			row = append(row, string(ch))
		}
		retval = append(retval, row)
	}
	return retval
}

func reorient(direction int) int {
	switch direction {
	case east:
		return south
	case south:
		return west
	case west:
		return north
	default:
		return east
	}
}

func getTick(direction int) string {
	switch direction {
	case east:
		return ">"
	case west:
		return "<"
	case north:
		return "^"
	default:
		return "v"
	}
}

func getNextPos(currPos Position, direction int) Position {
	switch direction {
	case east:
		currPos.Col += 1
	case west:
		currPos.Col -= 1
	case north:
		currPos.Row -= 1
	case south:
		currPos.Row += 1
	default:
		log.Fatalf("received unknown direction %d", direction)
	}
	return currPos
}

func move(positions *[][]string, direction int, currPos Position) *[]Position {
	orderedPositions := []Position{}
	movedPositions := map[Position]bool{}
	rLimit, cLimit := len(*positions), len((*positions)[0])
	nextPos := currPos

	for {
		if _, ok := movedPositions[currPos]; !ok {
			orderedPositions = append(orderedPositions, currPos)
		}
		movedPositions[currPos] = true
		nextPos = getNextPos(currPos, direction)
		if nextPos.Row >= rLimit || nextPos.Row < 0 ||
			nextPos.Col >= cLimit || nextPos.Col < 0 {
			break
		}
		if (*positions)[nextPos.Row][nextPos.Col] == "#" {
			direction = reorient(direction)
			continue
		}
		currPos = nextPos
	}
	return &orderedPositions
}

func moveLoop(positions *[][]string, currPos Position, size Position) bool {
	movedPositions := map[Position]bool{}
	nextPos := currPos
	for {
		if _, ok := movedPositions[currPos]; ok {
			return true
		} else {
			movedPositions[currPos] = true
		}
		nextPos = getNextPos(currPos, currPos.Value)
		nextPos.Value = currPos.Value
		if nextPos.Row >= size.Row || nextPos.Row < 0 ||
			nextPos.Col >= size.Col || nextPos.Col < 0 {
			return false
		}
		if (*positions)[nextPos.Row][nextPos.Col] == "#" {
			currPos.Value = reorient(nextPos.Value)
			continue
		}
		currPos = nextPos
	}
}

func getGuardPositionOrientation(data *[][]string) (Position, int) {
	for rowNum, row := range *(data) {
		for colNum, ch := range row {
			if ch == "^" {
				return Position{
					Row: rowNum,
					Col: colNum,
				}, north
			}
		}
	}
	log.Fatal("could not find guard!")
	return Position{}, -1
}

func copyGrid(g *[][]string) *[][]string {
	copiedGrid := [][]string{}

	for rNum := range len(*g) {
		row := []string{}
		for cNum := range len((*g)[0]) {
			row = append(row, (*g)[rNum][cNum])
		}
		copiedGrid = append(copiedGrid, row)
	}

	return &copiedGrid
}

func part01(data *[][]string) *[]Position {
	guardPos, orientation := getGuardPositionOrientation(data)
	positions := move(data, orientation, guardPos)
	fmt.Println("Part 01:", len(*positions))
	return positions
}

func part02(data *[][]string) {
	positions := part01(data)
	size := Position{Row: len(*data), Col: len((*data)[0])}
	guardPos, orientation := getGuardPositionOrientation(data)
	guardPos.Value = orientation
	numLoops := 0
	for _, pos := range (*positions)[1:] {
		(*data)[pos.Row][pos.Col] = "#"
		if moveLoop(data, guardPos, size) {
			numLoops++
		}
		(*data)[pos.Row][pos.Col] = "."
	}
	fmt.Println("Part 02:", numLoops)
}

func Run(dir string) {
	utils.PartPrinter("DAY 06")
	data := processInput(dir + "/day06")
	part02(&data)
}
